### ISSUE TITLE
Cousin ablation: implement EDL and −EML, compare against EML on recovery / size / stability

### ISSUE THREAD

**Issue by oaustegard** [oaustegard/eml-sr#36]
## Motivation

The paper establishes EML is not unique: §3 (p.9) names two cousins, §5 repeats and sharpens the claim.

| Operator | Formula | Distinguished constant | Why |
|---|---|---|---|
| `eml` | `exp(x) − ln(y)` | `1` | `ln(1) = 0`, identity of `−` |
| `edl` | `exp(x) / ln(y)` | `e` | `ln(e) = 1`, identity of `÷` |
| `−eml` | `ln(x) − exp(y)` | `−∞` | `exp(−∞) = 0`, identity of `−` |

All three claim the same completeness property (universal basis for elementary functions). The paper asserts this but doesn't run the side-by-side comparison on training dynamics, recovery rates, tree sizes, or numerical stability. The repo currently implements only `eml`.

Running the ablation answers: **are the cousins genuinely equivalent, or does EML win on some dimension the paper didn't measure?** Plausibly interesting outcomes:

- Recovery rate at depth `k` differs — some basins are easier to find.
- Canonical tree sizes differ. (EDL's multiplication might be cheaper than EML's depth-8 chain, since EDL already composes via `÷`.)
- Numerical stability differs under noise (relevant to the analog story in #34/#35). EDL's division may be worse than EML's subtraction at low `ln(y)` values.
- `−eml`'s `−∞` constant is numerically awful to carry around; this may be a disqualifier even if completeness holds.

## Depends on

- `#31` (EML compiler) — apples-to-apples comparison requires a compiler per operator, or a single compiler parameterized on operator choice. The latter is cleaner.

### A. Parameterize the engine on operator choice

Refactor `eml_sr.py` so `eml_op`, the canonical identities used by the simplifier, and the terminal constant are all swappable:

```python
class OperatorConfig:
    name: str                  # "eml" | "edl" | "neg_eml"
    op: Callable[[T, T], T]    # torch op
    terminal: complex          # 1, e, -inf
    identities: dict[str, Tree]  # exp, ln, e, pi, neg, add, mul, div, inv, pow
    
EML = OperatorConfig(name="eml", op=lambda x,y: torch.exp(x) - torch.log(y), terminal=1+0j, identities={...})
EDL = OperatorConfig(name="edl", op=lambda x,y: torch.exp(x) / torch.log(y), terminal=complex(math.e), identities={...})
NEG_EML = OperatorConfig(name="neg_eml", op=lambda x,y: torch.log(x) - torch.exp(y), terminal=complex(float("-inf")), identities={...})

def discover(X, y, ..., op=EML): ...
```

Minimum: `EMLTree1D` and `GrowingEMLTree` take an `op` argument; `Normalizer` and `discover`/`discover_curriculum` thread it through; simplifier reads from `op.identities`.

### B. Derive canonical identities per operator

Each operator needs its own identity table. For EML these already exist (CLAUDE.md §5). For the cousins, they need to be derived from scratch and verified numerically:

- **EDL** — `exp(x) = edl(x, e)` (since `ln(e) = 1`, so `edl(x, e) = exp(x)/1 = exp(x)`). `ln(x) = edl(e, edl(edl(e, x), e))` or similar — derive carefully.
- **−EML** — `ln(x) = −eml(x, −∞)` (since `exp(−∞) = 0`). `exp(x) = −eml(−eml(−eml(e, x), e), e)` or similar — derive.

Numerically verify every identity at sample points (e.g. `x = 0.577`, compare `derived_tree(x)` vs `math.log(x)`) before declaring it canonical.

### C. Benchmark: `benchmarks/cousin_ablation.py`

For each operator in `{EML, EDL, −EML}`, run the same target set at matched budget and record:

- Recovery rate per depth (the paper's §4.3 sweep: d2, d3, d4, d5, d6 from random init).
- Canonical tree sizes for a curated identity list (`exp`, `ln`, `1/x`, `x+y`, `x·y`, `x/y`, `sqrt(x)`).
- Wall time to first hit.
- Training stability: mean and stddev of final loss across seeds.
- Numerical failure rate: how often does training blow up (NaN, overflow) per operator?

Emit `benchmarks/cousin_ablation.md` with a side-by-side comparison table.

### D. Findings writeup

One-paragraph verdict per dimension:
- Which operator has the shortest canonical trees, and by how much?
- Which has the best recovery-from-random-init?
- Which is most numerically stable?
- Recommendation: should EML remain the default, or does a cousin win on a dimension that matters for a specific use case (e.g. analog, per #34)?

## Design decisions to call out

1. **−EML's `−∞` terminal.** Python `float("-inf")` propagates fine through `exp` (→ 0), but any subsequent `ln(0)` hits `−∞` again. Monitor for infinity-NaN cascades; the paper explicitly flags this in §4.1. May require tighter clamping or a rewrite that keeps `−∞` purely symbolic until the last step.
2. **EDL's `ln(y)` denominator.** Division by `ln(y)` blows up at `y = 1`. Input normalization must keep `y` away from 1 by some epsilon, or the training loop must clamp. Document explicitly.
3. **Complex intermediates.** All three operators inherit the complex-principal-branch machinery from EML. No change there — `complex128` throughout.
4. **Simplifier.** Currently tightly coupled to EML's identity table. Refactor to consume `op.identities` from the config; the rewriter logic itself is operator-agnostic (it walks the tree and applies rewrites from a dict).

## Non-goals

- Not claiming a winner up front. The point is to measure, not prescribe.
- Not porting the simplifier to a fully general term-rewriting system. Hardcoded identity tables per operator are fine for this issue.
- Not implementing the ternary operator. That's a separate issue (candidate follow-up) — the ternary is genuinely underspecified in the paper and needs its own VerifyBaseSet check first.
- Not integrating with PySR benchmark (#33). Each operator can be a separate row in that table once this issue lands.

## Paper references

- §3 (p.9) — "EML has at least two cousins: EDL defined as exp(x)/ln(y) with constant e, and −EML defined as ln(x) − exp(y) with constant negative infinity."
- §5 (p.15) — "EML is not unique; close variants such as EDL and the swapped-argument form eml(y,x) likely exhibit similar universality properties."
- `docs/odrzywolek-2026-eml.anchors.json` — filter by `section in {"3", "5"}` for the full claim set.
- Author's Rust `VerifyBaseSet` reference at `github.com/VA00/SymbolicRegressionPackage` — worth consulting for the canonical identities if the paper's text is ambiguous.

**Comment by oaustegard** [comment]
Depends on #31 (needs the compiler parameterized on operator choice for apples-to-apples comparison; at minimum, EMLTree1D should accept an `op` argument before this lands).
